### PR TITLE
Allow for cylindrical and spherical shell domains

### LINKED ISSUE #3115: Allow for cylindrical & spherical shell domains
### Description

At the moment we cannot simulate spherical shells with a boundary condition at the inner boundary other than no flux, as PyBaMM assumes the domain is always `[0, R]` and thus throws an error, see
https://github.com/pybamm-team/PyBaMM/blob/fffe650253596f14b6918d9e57fdd912bc3cfbdc/pybamm/discretisations/discretisation.py#L457-L464

Apparently it works for cylindrical polar (the error does not catch that case) but it makes me suspect that this feature is not well tested, so it would be good to also add some easy tests against analytical solutions to ensure this functionality works correctly.

### Motivation

Current models do not make use of this feature, but one can imagine how it could be relevant (e.g. core shell models).

In any case, for completeness of the functionality it would be good to enable this anyway.

### Possible Implementation

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by brosaplanella**
# Description

Now we only throw and error for non zero-Neumann bcs if the boundary is at `r = 0`, which means cylindrical and spherical shells can be solved with any boundary condition. Also added some integration tests to check it works.

Fixes #3115

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by brosaplanella** [comment]
This is now ready to review

File: tests/integration/test_spatial_methods/test_finite_volume.py [hunk 1]
```diff
@@ -294,6 +294,51 @@ def get_error(m):
...
+    model.initial_conditions = {u: pybamm.Scalar(0)}
+    model.variables = {"u": u, "r": r}
+    geometry = {"domain": {r: {"min": pybamm.Scalar(1), "max": pybamm.Scalar(2)}}}
+    submesh_types = {"domain": pybamm.Uniform1DSubMesh}
+    var_pts = {r: 500}
+    mesh = pybamm.Mesh(geometry, submesh_types, var_pts)
+    spatial_methods = {"domain": pybamm.FiniteVolume()}
+    disc = pybamm.Discretisation(mesh, spatial_methods)
+    disc.process_model(model)
+    solver = pybamm.CasadiAlgebraicSolver()
```
Comment (tinosulzer) [inline tests/integration/test_spatial_methods/test_finite_volume.py:318]: It's a bit strange to have a solver inside this test. Not the end of the world but would be better without, so that this is really a unit test (and not dependent on the solver). Is there another way to test the same thing?

**Comment by brosaplanella** [comment]
It's an integration test though, isn't it fine to have a solver there?

**Comment by tinosulzer** [comment]
Ah yes true, then it's fine

**Review by tinosulzer — APPROVED** [review]
